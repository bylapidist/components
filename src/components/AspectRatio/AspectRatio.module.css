.aspectratio {
    position: relative;
    overflow: hidden;
    height: 100%;
}

.aspectratio > :first-child {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
}

.aspectratio > :first-child > img {
    height: auto;
    object-fit: scale-down;
    object-position: 50% 50%;
}

.aspectratio:before {
    display: block;
    content: '';
}

.aspectratio-1-1:before {
    padding-bottom: calc(100% / (1 / 1));
}

.aspectratio-1-2:before {
    padding-bottom: calc(100% / (1 / 2));
}

.aspectratio-3-2:before {
    padding-bottom: calc(100% / (3 / 2));
}

.aspectratio-4-3:before {
    padding-bottom: calc(100% / (4 / 3));
}

.aspectratio-16-9:before {
    padding-bottom: calc(100% / (16 / 9));
}

.aspectratio-9-16:before {
    padding-bottom: calc(100% / (9 / 16));
}
