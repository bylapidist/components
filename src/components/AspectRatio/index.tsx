import * as React from 'react';
import cx from 'classnames';
import '@lapidist/styles/dist/css/light.css';
import type { BaseProps } from '../types';
import styles from './AspectRatio.module.css';

export type Ratio = '1/1' | '1/2' | '3/2' | '4/3' | '16/9' | '9/16';

export interface AspectRatioProps extends React.HTMLProps<HTMLDivElement> {
    readonly ratio: Ratio;
}

const AspectRatio = ({
    className,
    as: Component = 'div',
    testId = 'AspectRatio',
    ratio,
    children
}: BaseProps & AspectRatioProps) => (
    <Component
        className={cx(
            styles.aspectratio,
            styles[`aspectratio-${ratio.replace('/', '-')}`],
            className
        )}
        data-testid={testId}
    >
        <div>{children}</div>
    </Component>
);

AspectRatio.displayName = 'AspectRatio';
export default AspectRatio;
