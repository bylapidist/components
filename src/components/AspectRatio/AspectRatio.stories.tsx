import * as React from 'react';
import type { Meta, StoryObj, StoryFn } from '@storybook/react-vite';
import { BaseProps } from '../types';
import AspectRatio, { AspectRatioProps } from './index';

export default {
    title: 'Components/AspectRatio',
    component: AspectRatio
} as Meta;

const defaultArgs: BaseProps & AspectRatioProps = {
    as: 'div',
    testId: 'AspectRatio',
    ratio: '1/1'
};

const DefaultTemplate: StoryFn<BaseProps & AspectRatioProps> = (args) => (
    <AspectRatio {...args}>Hello world</AspectRatio>
);

const ImageTemplate: StoryFn<BaseProps & AspectRatioProps> = (args) => (
    <AspectRatio {...args}>
        <img src="https://placekitten.com/800/800" alt="A pretty cat" />
    </AspectRatio>
);

export const Default: StoryObj<BaseProps & AspectRatioProps> =
    DefaultTemplate.bind({});
Default.args = {
    ...defaultArgs
};

export const WithImage: StoryObj<BaseProps & AspectRatioProps> =
    ImageTemplate.bind({});
WithImage.args = {
    ...defaultArgs,
    ratio: '16/9'
};
