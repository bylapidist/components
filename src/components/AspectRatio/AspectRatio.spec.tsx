import * as React from 'react';
import { cleanup, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import AspectRatio, { Ratio } from './index';

const setup = (Component: React.ReactElement) => render(Component);

afterEach(cleanup);

const ratios: Ratio[] = ['1/1', '1/2', '3/2', '4/3', '16/9', '9/16'];

test.each(ratios)('it works with ratios', (ratio) => {
    const { container } = setup(
        <AspectRatio ratio={ratio}>Hello world</AspectRatio>
    );
    expect(container.firstChild).toMatchSnapshot();
});

test('it works with default testId', () => {
    setup(<AspectRatio ratio={ratios[0]}>Hello world</AspectRatio>);
    expect(screen.getByTestId('AspectRatio')).toBeTruthy();
});

test('it works with specified testId', () => {
    setup(
        <AspectRatio ratio={ratios[0]} testId={'TestId'}>
            Hello world
        </AspectRatio>
    );
    expect(screen.getByTestId('TestId')).toBeTruthy();
});

test('it works with as', () => {
    setup(
        <AspectRatio ratio={ratios[0]} as="span">
            Hello world
        </AspectRatio>
    );
    expect(screen.getByTestId('AspectRatio').nodeName).toBe('SPAN');
});

test('it works with className', () => {
    setup(
        <AspectRatio ratio={ratios[0]} className="test">
            Hello world
        </AspectRatio>
    );
    expect(
        screen.getByTestId('AspectRatio').classList.contains('test')
    ).toBeTruthy();
});
