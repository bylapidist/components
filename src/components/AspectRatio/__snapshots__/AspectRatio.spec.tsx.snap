// Vitest Snapshot v1

exports[`AspectRatio > it works with ratios 1`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-1-1_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`AspectRatio > it works with ratios 2`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-1-2_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`AspectRatio > it works with ratios 3`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-3-2_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`AspectRatio > it works with ratios 4`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-4-3_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`AspectRatio > it works with ratios 5`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-16-9_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`AspectRatio > it works with ratios 6`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-9-16_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 1`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-1-1_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 2`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-1-2_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 3`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-3-2_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 4`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-4-3_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 5`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-16-9_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;

exports[`it works with ratios 6`] = `
<div
  class="_aspectratio_779ae1 _aspectratio-9-16_779ae1"
  data-testid="AspectRatio"
>
  <div>
    Hello world
  </div>
</div>
`;
